components/push/src/crypto.rs

```diff
@@ -9,132 +9,41 @@
 //! This uses prime256v1 EC encryption that should come from internal crypto calls. The "application-services"
 //! module compiles openssl, however, so might be enough to tie into that.
 
-use log;
-use std::clone;
-use std::cmp;
-use std::fmt;
-
+use crate::error;
 use ece::{
-    Aes128GcmEceWebPushImpl, AesGcmEceWebPushImpl, AesGcmEncryptedBlock, LocalKeyPair,
-    LocalKeyPairImpl,
+    Aes128GcmEceWebPushImpl, AesGcmEceWebPushImpl, AesGcmEncryptedBlock, EcKeyComponents,
+    LocalKeyPair, LocalKeyPairImpl,
 };
-use openssl::ec::EcKey;
+use log;
 use openssl::rand::rand_bytes;
-
-use crate::error;
+use serde_derive::*;
 
 pub const SER_AUTH_LENGTH: usize = 16;
 pub type Decrypted = Vec<u8>;
 
-/// build the key off of the OpenSSL key implementation.
-/// Much of this is taken from rust_ece/crypto/openssl/lib.rs
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
 pub struct Key {
-    /// A "Key" contains the cryptographic Web Push Key data.
-    private: LocalKeyPairImpl,
-    pub public: Vec<u8>,
+    ec_key: EcKeyComponents,
     pub auth: Vec<u8>,
 }
 
-impl clone::Clone for Key {
-    fn clone(&self) -> Key {
-        Key {
-            private: LocalKeyPairImpl::new(&self.private.to_raw()).unwrap(),
-            public: self.public.clone(),
-            auth: self.auth.clone(),
-        }
+impl Key {
+    pub fn key_pair(&self) -> error::Result<LocalKeyPairImpl> {
+        LocalKeyPairImpl::from_raw_components(&self.ec_key).map_err(|e| {
+            error::ErrorKind::EncryptionError(format!(
+                "Could not re-create key from components: {:?}",
+                e
+            ))
+            .into()
+        })
     }
-}
 
-impl fmt::Debug for Key {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(
-            f,
-            "{{private: {:?}, public: {:?}, auth: {:?}}}",
-            base64::encode_config(&self.private.to_raw(), base64::URL_SAFE_NO_PAD),
-            base64::encode_config(&self.public, base64::URL_SAFE_NO_PAD),
-            base64::encode_config(&self.auth, base64::URL_SAFE_NO_PAD)
-        )
+    pub fn private_key(&self) -> &[u8] {
+        self.ec_key.private_key()
     }
-}
 
-impl cmp::PartialEq for Key {
-    fn eq(&self, other: &Key) -> bool {
-        self.private.to_raw() == other.private.to_raw()
-            && self.public == other.public
-            && self.auth == other.auth
-    }
-}
```

Comment:
Do we now get these extra traits for free, or do we no longer need them?

Reply:
We get them for free 👍